From 035a11cd712e809c85794bc48b33d1d24148538b Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Wed, 26 Feb 2025 18:34:14 +0100
Subject: [PATCH] serial: stm32: maintain console clock during suspend

To avoid many issues with bootloaders (that shares the control of the
console clock), we prefer to keep the clock gated when going into low
power modes.

Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I25142ff4a4f7b07ee6f37d876863592544fcb06e
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/440230
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/tty/serial/stm32-usart.c | 11 ++++++++++-
 1 file changed, 10 insertions(+), 1 deletion(-)

--- a/drivers/tty/serial/stm32-usart.c
+++ b/drivers/tty/serial/stm32-usart.c
@@ -2515,6 +2515,10 @@ static int __maybe_unused stm32_usart_se
 			stm32_usart_enable_am(stm32port);
 	}
 
+	/* Maintain the console clock to avoid issue with the bootloaders. */
+	if (uart_console(port))
+		clk_prepare_enable(stm32port->clk);
+
 	uart_suspend_port(&stm32_usart_driver, port);
 
 	/*
@@ -2550,7 +2554,12 @@ static int __maybe_unused stm32_usart_se
 			return ret;
 	}
 
-	return uart_resume_port(&stm32_usart_driver, port);
+	ret = uart_resume_port(&stm32_usart_driver, port);
+
+	if (uart_console(port))
+		clk_disable_unprepare(stm32port->clk);
+
+	return ret;
 }
 
 static int __maybe_unused stm32_usart_runtime_suspend(struct device *dev)
